AmplifyPlugins/API/Sources/AWSAPIPlugin/AppSyncRealTimeClient/AppSyncRealTimeClient.swift

```diff
@@ -0,0 +1,457 @@
+
+        Task { await self.subscribeToWebSocketEvent() }
+    }
+
+    deinit {
+        log.debug("Deinit AppSyncRealTimeClient")
+        subject.send(completion: .finished)
+        cancellables = Set()
+        cancellablesBindToConnection = Set()
+    }
+
+    /**
+     Connecting to remote AppSync real-time server.
+     */
+    func connect() async throws {
+        switch self.state.value {
+        case .connecting, .connected:
+            log.debug("[AppSyncRealTimeClient] client is already connecting or connected")
+            return
+        case .disconnecting:
+            try await waitForState(.disconnected)
+        case .connectionDropped, .disconnected, .none:
+            break
+        }
+
+        guard self.state.value != .connecting else {
+            log.debug("[AppSyncRealTimeClient] actor reentry, state has been changed to connecting")
+            return
+        }
+
+        self.state.send(.connecting)
+        log.debug("[AppSyncRealTimeClient] client start connecting")
+
+        try await RetryWithJitter.execute { [weak self] in
+            guard let self else { return }
+            try await AppSyncRealTimeRequest.sendRequest(
+                request: .connectionInit,
+                responseStream: subject.eraseToAnyPublisher()
+            ) { [weak self] _ in
+                await self?.webSocketClient.connect(
+                   autoConnectOnNetworkStatusChange: true,
+                   autoRetryOnConnectionFailure: true
+                )
+            }
+        }
+    }
+
+    /**
+     Disconnect only when there are no subscriptions exist.
+     */
+    func disconnectWhenIdel() async {
+        if self.subscriptions.isEmpty {
+            log.debug("[AppSyncRealTimeClient] no subscription exist, client is trying to disconnect")
+            await disconnect()
+        } else {
+            log.debug("[AppSyncRealTimeClient] client only try to disconnect when no subscriptions exist")
+        }
+    }
+
+    /**
+     Disconnect from AppSync real-time server.
+     */
+    func disconnect() async {
+        guard self.state.value != .disconnecting else {
+            log.debug("[AppSyncRealTimeClient] client already disconnecting")
+            return
+        }
+
+        defer { self.state.send(.disconnected) }
+
+        log.debug("[AppSyncRealTimeClient] client start disconnecting")
+        self.state.send(.disconnecting)
+        self.cancellablesBindToConnection = Set()
+        await self.webSocketClient.disconnect()
+        log.debug("[AppSyncRealTimeClient] client is disconnected")
+    }
+
+    /**
+     Subscribing to a query with unique identifier.
+     - Parameters:
+        - id: unique identifier
+        - query: GraphQL query for subscription
+
+     -  Returns:
+        A never fail data stream for AppSyncSubscriptionEvent.
+     */
+    func subscribe(id: String, query: String) async throws -> AnyPublisher<AppSyncSubscriptionEvent, Never> {
```

Comment:
is it typical to accept the `id` instead of generating it internally to the call? Would generating it internally and return it as ` (id: String, publisher: AnyPublisher<AppSyncSubscriptionEvent, Never>)` be better?

Reply:
We could generate it internally. Is there any benefit to do it internally? The subscriber need the id for unsubscribing. Handling the tuple returns is cumbersome than passing an exsisting one.

Reply:
discussed, since it is used internally, passing `id` simplifies the return type greatly